sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -649,6 +660,11 @@ private[joins] class SortMergeJoinScanner(
   // Initialization (note: do _not_ want to advance streamed here).
   advancedBufferedToRowWithNullFreeJoinKey()
 
+  def destruct(): Unit = {
+     while (streamedIter.advanceNext()) {}
+     while (bufferedIter.advanceNext()) {}
```

Comment:
It does introduce extra overhead. The other way is to introduce a new interface for RowIterator to destruct itself. But memory leak is worse than extra overhead, because it causes more spill.